Avoid redundant Pod restarts

jandubois opened the pull request:
Exported properties are no longer stored in secrets per pod, but in a secret per stateful set. It records the exported values. It also stores the `CONFIGGIN_VERSION_TAG`, which includes the chart version and the secrets generation, so it will change whenever there is a full restart of all pods.

Fissile will generate `CONFIGGIN_IMPORT_$ROLE` environment variables that reference the `CONFIGGIN_VERSION_TAG` in the secret of the exporting role. This prevents the importing pod from starting until the secret with the exported properties has been written.

This obviously requires that any mutual or cyclical imports between roles are resolved in the role manifest, either by using properties, or by patching some scripts to remove the bosh link and break the cycle. Otherwise there would be a deadlock as all roles in the cycle would wait forever for their imports to become ready.

There is no problem with cycles in imports from different jobs inside the same pod, as they are resolved simultaneously.

When properties are changed without a chart upgrade or secrets rotation, the pod templates of all importing stateful sets are annotated with the digest values of the exported values. The stateful set will automatically restart pods when the checksum of the pod spec has changed.

To avoid redundant restarts when the properties match the original deployment values, the annotations are not set (or removed if they already exist) when the digest matches the initial digest value.

andreas-kupries reviewed (approved):
Looks ok to me. Waiting for Mark and his checks.

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
+
+  # Write exported properties to secret and potentially restart affected pods.
+  def export_job_properties(jobs)
+    return unless instance_group == ENV["KUBERNETES_CONTAINER_NAME"]
```
We should have a comment about this being needed to handle sidecar containers.

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
+  # Write exported properties to secret and potentially restart affected pods.
+  def export_job_properties(jobs)
+    return unless instance_group == ENV["KUBERNETES_CONTAINER_NAME"]
+    return unless self_pod['metadata']['ownerReferences'][0]['kind'] == "StatefulSet"
```
And a comment on needing this because… something.

mook-as replied:
I think you mean `errands` here, but this is close enough.

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
+    secret.data ||= {}
+
+    version_tag = ENV["CONFIGGIN_VERSION_TAG"]
+    new_tag = !secret.data[version_tag]
```
```suggestion
    new_tag = !secret.data.has_key?(version_tag)
```

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
+      # order is already controlled by the "CONFIGGIN_IMPORT_#{role}" references to the new
+      # tags in the corresponding secrets. A missing import annotation reflects this set of
+      # exported properties because the sts as deployed by helm doesn't have any annotation
+      # and we don't want to trigger a redundant pod restart by adding them.
```
Hmm, I think my brain is broken this morning… I can't seem to understand that last sentence.

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
-      rescue
+      encoded_digest = Base64.encode64(digests[name])
+      # XXX do we actually need these?
+      # secret.data["skiff-exported-digest-#{name}"] = encoded_digest
```
No I don't think we need that

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
-    secret.data = {}
     digests = {}
     jobs.each do |name, job|
       digests[name] = property_digest(job.exported_properties)
```
Things will probably be clearer if we also keep this in a temporary variable.

mook-as replied:
We discussed this offline; I just got confused by the GitHub diff view.

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
+    version_tag = ENV["CONFIGGIN_VERSION_TAG"]
+    new_tag = !secret.data[version_tag]
+    if new_tag
+      secret.data = {version_tag => ""}
```
This will clobber any existing values (and that's intentional), right?

I think the ruby way is to spell this like:
```ruby
secret_data = {version_tag => ''} if new_tag # delete all existing secret values
```

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,89 @@ def generate_jobs(job_configs, templates)
+        secret.data["skiff-initial-digest-#{name}"] = encoded_digest
+      end
+      if secret.data["skiff-initial-digest-#{name}"] == encoded_digest
+        digests[name] = nil
```
So if the secret already has a matching digest (or if the tag is new), we'll delete the digest from the annotation?

jandubois replied:
Yes, that is what the comment above was trying to explain: There is no annotation for the property values exported by an install or update, so pods don't need to restart due to the template change.

mook-as commented on lib/kube_link_generator.rb:
```diff
@@ -86,43 +75,9 @@ def get_pods_for_role(role_name, job, options = {})
-    end
+  def get_exported_properties(role_name, job_name)
+    secret = client.get_secret(role_name, namespace)
+    JSON.parse(Base64.decode64(secret.data["skiff-exported-properties-#{job_name}"]))
```
This will blow up if the secret is missing right? (Assuming that's expected, document something in a comment?)

mook-as reviewed (changes requested):
Also need to fix up the tests of course.

Mostly here to prevent accidental merges.

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,86 @@ def generate_jobs(job_configs, templates)
+      # initial values because the helm chart doesn't include any annotations, and we don't
+      # want to trigger a pod restart by adding them.
+      if new_tag
+        secret.data["skiff-initial-digest-#{name}"] = encoded_digest
```
```suggestion
      secret.data["skiff-initial-digest-#{name}"] = encoded_digest if new_tag
```

jandubois replied:
I normally like using statement modifiers like this, but in this case I prefer to write the conditions first, to make it visually more obvious.

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,86 @@ def generate_jobs(job_configs, templates)
+      if new_tag
+        secret.data["skiff-initial-digest-#{name}"] = encoded_digest
+      end
+      if secret.data["skiff-initial-digest-#{name}"] == encoded_digest
```
```suggestion
      digests[name] = nil if secret.data["skiff-initial-digest-#{name}"] == encoded_digest
```

jandubois replied:
Same here

mook-as commented on lib/configgin.rb:
```diff
@@ -58,73 +57,86 @@ def generate_jobs(job_configs, templates)
+      # order is already controlled by the "CONFIGGIN_IMPORT_#{role}" references to the new
+      # tags in the corresponding secrets. There is no annotation when importing this set of
+      # initial values because the helm chart doesn't include any annotations, and we don't
+      # want to trigger a pod restart by adding them.
```
Okay, trying to rephrase this just to see if I understand this correctly (no need to take my wording):

> In order to ensure that we restart pods when their imported properties are changed, we set annotations on the importing StatefulSet containing digests of the imported properties.

> On initial deployment, the StatefulSets do not have those annotations (due to the helm deployment); to avoid unnecessarily restarting those initial pods, we explicitly do _not_ set annotations for the initial values (when we have a new version tag).  In that case, pod startup ordering is controlled by the `CONFIGGIN_IMPORT_#{role}` environment variable references to the secret.

> In order to keep track of whether this is a new version tag, we place the version tag in the generated secret.

jandubois replied:
> In order to keep track of whether this is a new version tag, we place the version tag in the generated secret.

We place the tag into the secret so fissile can create the `CONFIGGIN_IMPORT_role` variables that reference them, so kube will delay container creation until the secret exists.

And we use the version tag because during updates (or secrets rotation) the old tag already exists, but we don't want that version of the secrets, we have to wait until they are replace by the new ones.

jandubois commented:
I've bumped the version and created a new tag. Let's not merge this until the tests have been updated and the commit has a proper commit message!

I'll just build a test GEM from the PR branch.

mook-as commented:
I think we could use pre-release tags for this (if you don't like `update-stemcell.sh`); but if you've already published then meh.

jandubois committed:
Bump version to 0.20.0

[jsc#CAP-150]

jandubois committed:
Avoid unnecessary pod restarts

Exported properties are no longer stored in secrets per pod, but in
a secret per stateful set. It records the exported values. It also
stores the CONFIGGIN_VERSION_TAG, which includes the chart version
and the secrets generation, so it will change whenever there is a
full restart of all pods.

Fissile will generate CONFIGGIN_IMPORT_$ROLE environment variables
that reference the CONFIGGIN_VERSION_TAG in the secret of the exporting
role. This prevents the importing pod from starting until the secret
with the exported properties has been written.

This obviously requires that any mutual or cyclical imports between
roles are resolved in the role manifest, either by using properties,
or by patching some scripts to remove the bosh link and break the
cycle. Otherwise there would be a deadlock as all roles in the cycle
would wait forever for their imports to become ready.

There is no problem with cycles in imports from different jobs inside
the same pod, as they are resolved simultaneously.

When properties are changed without a chart upgrade or secrets rotation,
the pod templates of all importing stateful sets are annotated with
the digest values of the exported values. The stateful set will
automatically restart pods when the checksum of the pod spec has changed.

To avoid redundant restarts when the properties match the original deployment
values, the annotations are not set (or removed if they already exist)
when the digest matches the initial digest value.

[jsc#CAP-150]

jandubois committed:
Allow service account override for configgin

If CONFIGGIN_SA_TOKEN is defined, use that value instead of the
pod service account for configgin. This is preferable so that the
rest of the pod does not have access to all the secrets in the
namespace.

The token may expire, but configgin only runs briefly when the pod
is created, so it should not be a problem that the token cannot be
refreshed.

[jsc#CAP-828]

jandubois commented:
I've fixed up the tests and updated comments some more since the reviews.

There are also bug fixes, including bumping `kubeclient` to 4.3 to get support for JSON merge patches and also https://github.com/cloudfoundry-incubator/configgin/commit/7aa29f2647cf19df2ce36a829bd7f3545d38ed05.

All changes have been squashed, so the history does not include the experimental tags 0.19.2 to 0.19.6 anymore.

I've also included the trivial change in https://github.com/cloudfoundry-incubator/configgin/pull/108/commits/c8e110be357fde52de6ab8b96ab0ea6e2d692e32 to avoid having to update configgin again if we want to address CAP-828 in the future.

mook-as commented on lib/configgin.rb:
```diff
@@ -133,15 +141,30 @@ def restart_affected_pods(expected_annotations)
+      sts.spec.template.metadata.annotations.each_pair do |key, value|
+        annotations[key] = value
+      end
+      digests.each_pair do |key, value|
```
I believe that's the same as `annotations.merge! digests`.  We can probably do that with `sts.spec.template.metadata.annotations` too, because `annotations` is definitely a plain hash.

andreas-kupries reviewed (approved):
LGTM.

mook-as commented on spec/lib/configgin_spec.rb:
```diff
@@ -76,70 +75,51 @@
+        statefulset = client.get_stateful_set('debugger', 'the-namespace')
+        expect(statefulset).not_to be_nil
+        expected_digest = statefulset.spec.template.metadata.annotations['skiff-in-props-instance-group-loggregator_agent']
+        expect(expected_digest).to eq property_digest(JSON.parse(exported_properties))
```
The naming's backwards here, no? (The digest we fetched is the actual digest, not the expected.)

mook-as commented on spec/spec_helper.rb:
```diff
@@ -32,6 +40,14 @@ def _create_single(type, resource)
+    object = _get_single(type, resource.metadata.name, resource.metadata.namespace)
+    raise %(Could not find #{type} "#{resource.metadata.namespace}/#{resource.metadata.name}" to patch) unless object
+
+    object.delete_if { true }
```
```suggestion
    object.clear
```

mook-as commented on spec/spec_helper.rb:
```diff
@@ -52,7 +68,7 @@ def method_missing(name, *args)
       raise "Don't know how to patch multiple #{type}s: #{name}" if name.to_s.end_with? 's'
 
-      return _patch_single(type, *args)
+      return _patch_single(type, "strategic", *args)
```
Nit: rubocop _should_ complain about using single quotes if there's no substitution.

mook-as reviewed (approved):
LGTM, tiny nits that I don't know if we should bother with.
